python/pyarrow/_exec_plan.pyx

```diff
@@ -61,10 +65,18 @@ cdef execplan(inputs, output_type, vector[CDeclaration] plan, c_bool use_threads
         CTable* c_table
         shared_ptr[CTable] c_out_table
         shared_ptr[CSourceNodeOptions] c_sourceopts
+        shared_ptr[CScanNodeOptions] c_scanopts
+        shared_ptr[CExecNodeOptions] c_input_node_opts
         shared_ptr[CSinkNodeOptions] c_sinkopts
         shared_ptr[CAsyncExecBatchGenerator] c_async_exec_batch_gen
         shared_ptr[CRecordBatchReader] c_recordbatchreader
         vector[CDeclaration].iterator plan_iter
+        vector[CDeclaration.Input] no_c_inputs
+
+    global _dataset_support_initialised
+    if not _dataset_support_initialised:
+        Initialize()  # Initialise support for Datasets in ExecPlan
+        _dataset_support_initialised = True
```

Comment:
Is there a reason to not do this at the module level? Is it costly in case the user would not end up using datasets?

Reply:
Good point, I'll change that to be at module level.